### PR TITLE
Add Wikilink titles support

### LINKED ISSUE #136: Support wiki links titles
As mentioned in docs:

> By default Marksman uses a document title's slug when referencing a document, however there is an configuration setting to use a file name or a file path instead.

and in [this reply](https://github.com/artempyanykh/marksman/issues/122#issuecomment-1348312210), like in Obsidian and Foam it could be really useful to add a `|` for a title/alias to the slug or filename/path:

`[[destination|title]]`

Example:
`Check my [[short-bio|bio]]`

Preview:
`Check my bio`

### PR DESCRIPTION
**Pull request by eguiraud**
With this patch wikilinks of the form [[doc#heading|title]] are correctly recognized as links to doc#heading.
Also go-to-definition works as normal.
Fixes #136.

Admittedly, this could have been done with a smaller patch, but the addition of the title-handling logic was making the function quite hard to follow (already before the patch it had 6+ levels of indentation).
So I took this chance to refactor the parsing to use a simple "functional finite state machine pattern". Basically we have a `parse()` function that kicks off the parsing and then transitions to other functions as needed via pattern matching (`parseTitle`, `parseHeading`, `parseDoc`, `parseEnd`).

- `parseDoc` can transition to all other functions but not to `parse` (we don't go back to the beginning) or fail early
- `parseHeading` can only transition to `parseTitle` or `parseEnd` (we don't go back to the doc) or fail early
- `parseTitle` can only transition to `parseEnd` or fail early
- `parseEnd` can either return success (this was a well-formed wiki link) or failure (not a well-formed wiki link)

I hope the new logic is easy to follow!

### CODE REVIEW

**Comment by artempyanykh** [comment]
> So I took this chance to refactor the parsing to use a simple "functional finite state machine pattern"

Love it, thanks! 

I'll have a closer look shortly. Meanwhile, could you please
1. run `make fmt` and `make check` to unbork the CI, and
2. add a test or two in [ParserTests.fs](https://github.com/artempyanykh/marksman/blob/main/Tests/ParserTests.fs) to exercise the new code path.

**Comment by eguiraud** [comment]
I applied the suggestions from `make fmt` and `make check`.

I need to spend some time understanding how the tests are set up, I'll try to do so soon.

Unrelated: while testing these changes interactively I found a problem that was not caught by the tests when suggesting completions for an empty `[[]]` wiki link, maybe there is a missing test case.

File: Marksman/Parser.fs [hunk 1]
```diff
@@ -63,105 +67,122 @@ module Markdown =
...
+            let isALink = parse ()
+            slice_ <- slice // update output parameter to modified slice state
 
-                let mutable offsetHashDelim: option<int> = None
-                let mutable found = false
-                let mutable current = slice.NextChar()
+            if isALink then
+                let offsetEnd = processor.GetSourcePosition(slice.Start)
+                let text = slice.Text.Substring(start, offsetEnd - offsetStart + 1)
+                let sliceOffset = offsetStart - start
```
Comment (artempyanykh) [inline Marksman/Parser.fs:166]: nit: maybe do the following inside `contentAndSpan` instead (to make the calculations a bit clearer):
```
let contentSliceStart = start + (span.Start - offsetStart)
let contentSliceLen = span.End - span.Start + 1
let content = slice.Text.Substring(contentSliceStart, contentSliceLen)
...
```

**Review by artempyanykh — COMMENTED** [review]
A couple nits, otherwise looks great! @eguiraud LMK if you need some help with testing.

File: Marksman/Parser.fs [hunk 2]
```diff
@@ -63,105 +67,122 @@ module Markdown =
...
-        override this.Match(processor, slice) =
-            let nextChar = slice.PeekCharExtra(1)
+        override this.Match(processor, slice_) =
+            let mutable docSpan: option<SourceSpan> = None
+            let mutable headingSpan: option<SourceSpan> = None
+            let mutable titleSpan: option<SourceSpan> = None
+
+            let mutable slice = slice_ // this copy is necessary because byrefs cannot be captured
+
+            // helper functions for finite state machine
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply Marksman/Parser.fs:83]: I changed the function name but I could not find a way to make the compiler happy with the mutating iteration, sorry :/

File: Marksman/Parser.fs [hunk 3]
```diff
@@ -63,105 +67,122 @@ module Markdown =
...
+            let withEnd offset (span: option<SourceSpan>) =
+                span |> Option.map (fun s -> SourceSpan(s.Start, offset - 1))
+
+            let notEscaped () = slice.PeekCharExtra(-1) <> '\\'
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply Marksman/Parser.fs:87]: done

File: Marksman/Parser.fs [hunk 4]
```diff
@@ -63,105 +67,122 @@ module Markdown =
...
+                    | _ -> parseDoc offset
 
-            if isRef then
-                let start = slice.Start
-                let offsetStart = processor.GetSourcePosition(start)
-                let offsetInnerStart = offsetStart + 2
+            // do the parsing (run the finite state machine)
+            let start = slice.Start
+            let offsetStart = processor.GetSourcePosition(start)
+            let isALink = parse ()
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply Marksman/Parser.fs:160]: done

File: Marksman/Parser.fs (hunk 1 above)
Comment (eguiraud) [reply Marksman/Parser.fs:166]: done

**Comment by eguiraud** [comment]
Applied suggestions from code review and added some tests, although I still don't understand how the tests works :sweat_smile: first time I write F#, sorry

File: Marksman/Parser.fs (hunk 2 above)
[the remark this answers is not in the document]
Comment (artempyanykh) [reply Marksman/Parser.fs:83]: No probs! Looks good!

**Comment by artempyanykh** [comment]
Great job, thanks @eguiraud! 

Re: tests -- I use snapshot tests quite often (also known as *golden testing*) where the test output is compared to a previously recorded output. I use https://github.com/theramis/Snapper library, which works albeit a bit fiddly wrt. updating the snapshots.

**Comment by tjex** [comment]
It doesn't seem to be working for me :(
updated via Mason in neovim to latest and tried to make two fresh links following the wikilink syntax. 
you can see by the lack of lsp error that marksman does see the note at `[[open source]]`.

the cwd of neovim is the same folder where the `[[open source]]` note is as well as where the test note is (being used to test the links below)

Screen Shot 2023-06-22 at 14 24 22

**Comment by artempyanykh** [comment]
@tjex it hasn't been released yet, just merged into main.

I'll cut a new release in a couple days. Otherwise, you may want to try building from source.
